File: nn/rnn/RNNCell.cpp
#include "RNNCell.h"

evo::nn::rnn::RNNCell::RNNCell(int hidden_size, int training_iterations, std::string hidden_activation, std::vector<float> weight_params,
	std::vector<float> bias_params, bool bias_bool, std::string fp)
: AbstractRNN::AbstractRNN(0, hidden_size, "", hidden_activation, weight_params, bias_params, bias_bool, training_iterations, 
	alpha, fp), I(hidden_size, 0, "", weight_params, bias_params, 0), H(hidden_size, hidden_size, "", weight_params, bias_params, bias_bool){

}
evo::nn::rnn::RNNCell::RNNCell(int hidden_size, int input_size, int training_iterations, std::string hidden_activation, std::vector<float> weight_params,
	std::vector<float> bias_params, bool bias_bool, std::string fp)
	: AbstractRNN::AbstractRNN(0, hidden_size, "", hidden_activation, weight_params, bias_params, bias_bool, training_iterations,
		alpha, fp), I(hidden_size, input_size, "", weight_params, bias_params, 0), H(hidden_size, hidden_size, "", weight_params, bias_params, bias_bool) {
	if (fp == "") {
		initialize_weights();
	}
}

void evo::nn::rnn::RNNCell::initialize_weights() {
	I.initialize_layer();
	H.initialize_layer();
	dh_dW1 = evo::empty(in_size, out_size);
	dh_dWh = evo::empty(out_size, out_size);
	dh_db = evo::empty(out_size);
}

void evo::nn::rnn::RNNCell::train(std::vector<float> error_vect) {
	int t = 1;
	while (t < timestamp) {
		int max_timestamp = m_min(t + iterations, timestamp);
		while (t < max_timestamp) {
			for (int j = 0; j < out_size; j++) {
				std::vector<float> hp = vectors["h" + std::to_string(t - 1)];
				float h_prv = hp[j];
				float activation_error = act.get_error("h" + std::to_string(t), j);
				float error = error_vect[j];

				H.modify_bias(j, error * dB1(activation_error, H.get_weight(j, j), dh_db[j], j));

#pragma omp parallel
				for (int i = 0; i < in_size; i++) {
					float x = vectors["x" + std::to_string(t)][i];

					I.modify_weight(i, j, error * dW1(activation_error, x, H.get_weight(i, j), dh_dW1[i][j], i, j));
				}

#pragma omp parallel
				for (int j2 = 0; j2 < out_size; j2++) {

					H.modify_weight(j2, j, error * dWh(activation_error, h_prv, H.get_weight(j2, j), dh_dWh[j2][j], j2, j));
				}
			}
			t += 1;
		}
		reset();
	}
}

float evo::nn::rnn::RNNCell::dW1(float ae, float x, float wh, float dh_prv, int i, int j) {
	float er = (x + wh * dh_prv) * ae;
	dh_dW1[i][j] = er;
	return er;
}

float evo::nn::rnn::RNNCell::dWh(float ae, float h_prv, float wh, float dh_prv, int i, int j) {
	float er = ae * (h_prv + wh * dh_prv);
	dh_dWh[i][j] = er;
	return er;
}

float evo::nn::rnn::RNNCell::dB1(float ae, float wh, float dh_prv, int j) {
	float er = ae * (wh * dh_prv + 1);
	dh_db[j] = er;
	return er;
}


void evo::nn::rnn::RNNCell::reset() {
	dh_dW1 = evo::empty(in_size, out_size);
	dh_dWh = evo::empty(out_size, out_size);
	dh_db = evo::empty(out_size);
}

std::vector<float> evo::nn::rnn::RNNCell::feed_forward_one_vect(std::vector<float> x) {
	std::string key = "h" + std::to_string(timestamp-1);
	std::vector<float> h1 = I.feedforward(x);
	std::vector<float> h2 = H.feedforward(vectors[key]);
	std::vector<float> h = evo::add(h1, h2);
	std::vector<float> y = act.activate(h, hidden_activation, ("h" + std::to_string(timestamp)));
	vectors["x" + std::to_string(timestamp)] = x;
	vectors["h" + std::to_string(timestamp)] = y;
	timestamp += 1;
	return y;
}


std::vector<float> evo::nn::rnn::RNNCell::feedforward(mtx v) {
	return AbstractRNN::feedforward(v);
}

template<typename T>
void evo::nn::rnn::RNNCell::print_vector(std::vector<T> a) {
	std::cout << "{ ";
	for (T e : a) {
		std::cout << e << " ";
	}
	std::cout << " }" << std::endl;
}

File: nn/rnn/RNNCell.py
from nn.Dense import Dense
import numpy as np
from nn.rnn.AbstractRNN import AbstractRNN


class RNNCell(AbstractRNN):
    """Vectors: all hidden and cell state vectors"""

    def __init__(self, hidden_size, input_size=None, training_iterations=5, hidden_activation='softmax', weight_param=(-1, 1)
                 , bias_params=(-1, 1), bias_bool=True, fp=''):
        super().__init__(input_size, hidden_size,  None, hidden_activation, (0, 5), (-1, 1), True, training_iterations)

        self.I = Dense(hidden_size, input_size, None, weight_param, bias_params, False)
        self.H = Dense(hidden_size, hidden_size, None, weight_param, bias_params, bias_bool)

        self.dh_dW1 = np.zeros(shape=(input_size, hidden_size))
        self.dh_dWh = np.zeros(shape=(hidden_size, hidden_size))
        self.dh_db = np.zeros(hidden_size)

        if fp == '':
            self.initiate_weights()

    def initiate_weights(self):
        self.I.initialize(), self.H.initialize()

    def feed_forward_one_vect(self, input_vect):
        h1 = self.I.feed_forward(input_vect)
        h2 = self.H.feed_forward(self.Vectors['h'+str(self.timestamp-1)])
        h = np.add(h1, h2)
        y = self.A.activation_function(h, self.hidden_activation, "a"+str(self.timestamp))

        self.Vectors['x' + str(self.timestamp)] = input_vect
        self.Vectors['h'+str(self.timestamp)] = y
        self.timestamp += 1
        return y

    def train(self, error_vect):
        timestamp = 1
        while timestamp < self.timestamp:
            max_timestamp = min(timestamp + self.iterations, self.timestamp)
            while timestamp < max_timestamp:
                for j in range(self.hidden_size):
                    h_prv = self.Vectors['h' + str(timestamp - 1)][j]
                    activation_error = self.A.errors["a" + str(timestamp)][j]
                    error = error_vect[j]
                    self.H.Bias[j] += error * self.dB1(activation_error, self.H.Weight[j][j], self.dh_db[j], j)
                    for i in range(self.input_size):
                        x = self.Vectors['x' + str(timestamp)][i]
                        self.I.Weight[i][j] += error * self.dW1(activation_error, x, self.H.Weight[j][j],
                                                                self.dh_dW1[i][j], i, j)

                    for j2 in range(self.hidden_size):
                        self.H.Weight[j2][j] += error * self.dWh(activation_error, h_prv, self.H.Weight[j2][j],
                                                                 self.dh_dWh[j2][j], j2, j)

                timestamp += 1
            self.reset()

    def dW1(self, ae, x, wh, dh_prv, i, j):
        er = (x + wh*dh_prv)*ae
        self.dh_dW1[i][j] = er
        return er

    def dWh(self, ae, h_prv, wh, dh_prv, i, j):
        er = ae*(h_prv + wh*dh_prv)
        self.dh_dWh[i][j] = er
        return er

    def dB1(self, ae, wh, dh_prv, j):
        er = ae*(wh*dh_prv + 1)
        self.dh_db[j] = er
        return er

    def dNext(self, error, ae, wh, dh_prv, i, j):
        er = (self.I.Weight[i][j] + wh * dh_prv) * ae * error
        return er

    def reset(self):
        self.dh_dW1 = np.zeros(shape=(self.I.input_size, self.hidden_size))
        self.dh_dWh = np.zeros(shape=(self.hidden_size, self.hidden_size))
        self.dh_db = np.zeros(self.hidden_size)

    def get_output(self):
        return self.Vectors['h'+str(self.timestamp-1)]


def test():
    s, t = 0, 0
    for _ in range(100):
        i, o = 20, 10
        R = RNNCell(o, i, weight_param=(0, 1), bias_params=(0, 1), hidden_activation='softmax')
        vects = [np.random.random(size=i) for j in range(i)]

        y = np.zeros(o)
        y[1] = 1

        # R.feed_forward(vects[0])

        for vect in vects:
            R.feed_forward(vect)

        for i in range(20):
            error_vect = np.subtract(y, R.get_output())
            R.train(np.multiply(error_vect, 0.06))

            for vect in vects:
                R.feed_forward(vect)

        t += 1
        if np.argmax(R.get_output()) == 1:
            s += 1

    return s / t


if __name__ == '__main__':
    print(test())
    # s, t = 0, 1
    #
    # i, o = 5, 5
    # R = RNNCell(i, o)
    # vects = [np.random.random(size=i) for j in range(4)]
    #
    # y = np.zeros(o)
    # y[1] = 1
    #
    # for vect in vects:
    #     R.feed_forward(vect)
    # print(R.Vectors.keys())
    # print(R.timestamp)
    # for _ in range(20):
    #     error_vect = np.subtract(y, R.get_output())
    #     R.gradient(np.multiply(error_vect, 0.05))
    #     for vect in vects:
    #         R.feed_forward(vect)
    #     # print(R.get_output())
    #
    # print(R.Vectors.keys())
    # print(R.timestamp)
    # print(R.dh_db)
    # print(R.I.Vectors.keys())
